tests/python/test_random.py

```diff
@@ -109,8 +109,19 @@ def gen(i: ti.i32):
 
 @ti.test(arch=[ti.cpu, ti.cuda])
 def test_random_f64():
+    '''
+    Tests the generation of well-behaved Float64 random numbers by granularity,
+    see #2251 for explanation.
+    '''
+    import numpy as np
+    n = int(2**23)
+    x = ti.field(ti.f64, shape=n)
+
     @ti.kernel
     def foo() -> ti.f64:
-        return ti.random(dtype=ti.f64)
+        for i in x:
+            x[i] = ti.random(dtype=ti.f64)
 
     foo()
+    diff = np.diff(np.sort(x.to_numpy()), 1)
+    assert np.min(diff) > 0
```

Comment:
It seems that we are assuming no two random numbers generated to have the same value. Not sure how safe it is to make this assumption. I guess a more robust way to test the granularity is to use something like 

```python
abs(np.min(np.diff(np.unique(np.sort(x)), 1)) - 1.0 / 18446744073709551616.0) < 1e-30
```
?

According to the birthday paradox, this holds with very high probability (but not sure with pseudo-random number generators...)

Reply:
Yes, I agree that results of unit tests should be made as deterministic as possible, and assuming no two random numbers collide indeed has a finite probability of failure. We can calculate the probability of "birthday problem" for 2^23 float64 numbers:

> The collision probability for the birthday problem with _d_ choices and sample size of _n_ [is approximately 1 - exp(-n^2/(2d))](https://en.wikipedia.org/wiki/Birthday_problem#Approximations).
Given float64 numbers between 0 and 1 has at least 2^52 values (upper bound estimate by ignoring the exponent bits), The collision probability is approximately 1 - exp(-2^(-7)) = 0.008.

Given the massive number of tests run in a CI environment, I guess a nearly 1% failure probability is indeed not acceptable. `abs(np.min(np.diff(np.unique(np.sort(x)), 1)) - 1.0 / 18446744073709551616.0) < 1e-30` assumes that two adjacent float64 numbers should be found in 2^23 samples, which has the same probability as two numbers collide and is very unlikely to pass.

Let's use a more straightforward solution:
```python
frac, _ = np.modf(x.to_numpy() * 4294967296)
assert np.min(frac) > 0
```
Our previous float32 casting implementation should always give integers after multiplying 2^32, but the full float64 implementation should not.

Reply:
OK sounds good! I like simple solutions :-) You may need to exclude values that happen to be a multiple of `1 / 4294967296`.

Reply:
Yes, the updated commit used `assert np.max(frac) > 0` so that the test will pass even if some values happen to be a multiple of 1/2^32. Conversely, a float32 random number is always a multiple of 1/2^32.